src/iosrv/ult.c

```diff
@@ -0,0 +1,538 @@
+ *
+ * \param[in] ops		All dss_collective ops to work on streams
+ *				include \a func(\a arg) for collective on all
+ *				server xstreams.
+ * \param[in] args		All arguments required for dss_collective
+ *				including func args.
+ * \param[in] flag		collective flag, reserved for future usage.
+ *
+ * \return			number of failed xstreams or error code
+ */
+int
+dss_task_collective_reduce(struct dss_coll_ops *ops,
+			   struct dss_coll_args *args, int flag)
+{
+	return dss_collective_reduce_internal(ops, args, false, flag);
+}
+
+/**
+ * General case:
+ * Execute \a ULT(\a arg) collectively on all server xstreams. Can only be
+ * called by ULTs. Can only execute tasklet-compatible functions. User specified
+ * reduction functions for aggregation after collective
+ *
+ * \param[in] ops		All dss_collective ops to work on streams
+ *				include \a func(\a arg) for collective on all
+ *				server xstreams.
+ * \param[in] args		All arguments required for dss_collective
+ *				including func args.
+ * \param[in] flag		collective flag, reserved for future usage.
+ *
+ * \return			number of failed xstreams or error code
+ */
+int
+dss_thread_collective_reduce(struct dss_coll_ops *ops,
+			     struct dss_coll_args *args, int flag)
+{
+	return dss_collective_reduce_internal(ops, args, true, flag);
+}
+
+static int
+dss_collective_internal(int (*func)(void *), void *arg, bool thread, int flag)
+{
+	int				rc;
+	struct dss_coll_ops		coll_ops = { 0 };
+	struct dss_coll_args		coll_args = { 0 };
+
+	coll_ops.co_func	= func;
+	coll_args.ca_func_args	= arg;
+
+	if (thread)
+		rc = dss_thread_collective_reduce(&coll_ops, &coll_args, flag);
+	else
+		rc = dss_task_collective_reduce(&coll_ops, &coll_args, flag);
+
+	return rc;
+}
+
+
+/**
+ * Execute \a func(\a arg) collectively on all server xstreams. Can only be
+ * called by ULTs. Can only execute tasklet-compatible functions.
+ *
+ * \param[in] func	function to be executed
+ * \param[in] arg	argument to be passed to \a func
+ * \param[in] flag	collective flag, reserved for future usage.
+ *
+ * \return		number of failed xstreams or error code
+ */
+int
+dss_task_collective(int (*func)(void *), void *arg, int flag)
+{
+	return dss_collective_internal(func, arg, false, flag);
+}
+
+/**
+ * Execute \a func(\a arg) collectively on all server xstreams. Can only be
+ * called by ULTs. Can only execute tasklet-compatible functions.
+ *
+ * \param[in] func	function to be executed
+ * \param[in] arg	argument to be passed to \a func
+ * \param[in] flag	collective flag, reserved for future usage.
+ *
+ * \return		number of failed xstreams or error code
+ */
+
+int
+dss_thread_collective(int (*func)(void *), void *arg, int flag)
+{
+	return dss_collective_internal(func, arg, true, flag);
+}
+
+/* ============== ULT create functions =================================== */
+
+static inline int
+sched_ult2pool(int ult_type)
+{
+	switch (ult_type) {
+	case DSS_ULT_DTX_RESYNC:
```

Comment:
seems we ever discussed long time ago that some tasks like dtx resync needs to be scheduled with highest priority, to avoid the case that there is lots of IO tasks and the dtx_resync cannot be ensured to schedule in time.
This patch removed the URGENT pool, not sure above is still a concern?

Reply:
First of all, I think we should define ABT pools by the type of ULTs, because the priority of ULT could change from time to time, instead of moving ULTs between different ABT pools with fixed priority, dynamically throttling certain ABT pools could be a better choice.

I think putting it in IO pool should be good enough, if later we find out that's not enough, we may introduce a new pool for it, what do you think?

Reply:
looks fine to me, thx.